### DISCUSSION TITLE
How to Keep Backend and Frontend Running for Parallel Cypress E2E Tests?

### DISCUSSION THREAD

**Discussion by mah-noor1** [cypress-io/cypress#30369] [Questions and Help]
I'm working on setting up a Cypress end-to-end (E2E) pipeline. My current setup involves starting backend services and the local frontend before running the Cypress tests.

The issue I'm facing is that I want to run the setup only once, keeping the backend services and frontend running in the background while executing the Cypress E2E tests in parallel. However, after successfully setting up the services, Cypress tests cannot connect to them.

Has anyone dealt with a similar issue or had any suggestions on achieving this setup?

![image](https://github.com/user-attachments/assets/2000f1f7-0bd5-44f1-80d2-3d63588a56f8)

**Comment by portyu9** [comment]
Hi @mah-noor1 

Cypress parallelization does **not** make multiple CI workers share `localhost` or background processes.

If you start your backend/frontend in a separate setup job, those processes normally exist only inside that job's VM/container. A Cypress worker running on another machine cannot reach them through `localhost`.

You generally have two options:

1. **Start the services in every Cypress worker** and wait for them to become ready before starting Cypress:

```sh
npm run start:backend &
npm run start:frontend &

npx wait-on http://127.0.0.1:3000 http://127.0.0.1:8080/health

npx cypress run --record --parallel
```

2. **Start the services once on a shared, network-reachable environment** and configure every Cypress worker to use that hostname/URL instead of `localhost`.

The Cypress Cloud parallel run/build ID coordinates which specs are assigned to workers; it does not share processes or networking between those workers.

Also, prefer an actual readiness check such as `wait-on` or a health endpoint rather than starting the processes in the background and immediately launching Cypress.

No screenshots or attachments were included in this discussion.

Please mark this answer as accepted if it helped, thank you.
